pkg/cmd/kubeconfig.go

```diff
@@ -45,21 +49,108 @@ func kubeconfigCmd(rootFlags *pflag.FlagSet) *cobra.Command {
 				return err
 			}
 
-			return runKubeconfig(gopts)
+			st, err := gopts.BuildState()
+			if err != nil {
+				return err
+			}
+
+			konfig, err := kubeconfig.Download(st)
+			if err != nil {
+				return err
+			}
+			fmt.Println(string(konfig))
+
+			return nil
 		},
 	}
 
+	cmd.AddCommand(kubeconfigGenerateCmd(rootFlags))
+
 	return cmd
 }
 
-// runKubeconfig downloads kubeconfig file
-func runKubeconfig(opts *globalOptions) error {
-	s, err := opts.BuildState()
-	if err != nil {
-		return err
+type kubeconfigGenerateOpts struct {
+	globalOptions
+	CommonName        string        `longflag:"cn"`
+	OrganizationNames []string      `longflag:"on"`
+	ShortSuperAdmin   bool          `longflag:"super-admin" shortflag:"s"`
+	TTL               time.Duration `longflag:"ttl"`
+}
+
+func kubeconfigGenerateCmd(rootFlags *pflag.FlagSet) *cobra.Command {
+	opts := &kubeconfigGenerateOpts{}
+
+	cmd := &cobra.Command{
+		Use:     "generate",
+		Short:   "Generate kubeconfig",
+		Long:    "Generate kubeconfig with given certificate parameters.",
+		Example: `kubeone kubeconfig generate -m mycluster.yaml -t tf.json --super-admin`,
+		RunE: func(_ *cobra.Command, _ []string) error {
+			gopts, err := persistentGlobalOptions(rootFlags)
+			if err != nil {
+				return err
+			}
+			opts.globalOptions = *gopts
+
+			st, err := gopts.BuildState()
+			if err != nil {
+				return err
+			}
+
+			return runKubeconfigGenerate(st, opts)
+		},
+	}
+
+	cmd.Flags().StringVar(
+		&opts.CommonName,
+		longFlagName(opts, "CommonName"),
+		"",
+		"CommonName (CN) for the generated client certificate.")
+
+	cmd.Flags().StringArrayVar(
+		&opts.OrganizationNames,
+		longFlagName(opts, "OrganizationNames"),
+		[]string{},
+		"OrganizationName (ON) for the generated client certificate.")
+
+	cmd.Flags().BoolVarP(
+		&opts.ShortSuperAdmin,
```

Comment:
We should have validation to reject `--super-admin` if ON and CN are provided.

Reply:
ON is getting (silently) overwritten and that might be a problem, but what's the problem with CN?